File: src/main/java/ru/uris/PType.java
package ru.uris;

public enum PType {
    STRING(0),
    DOUBLE(1),
    LONG(2),
    INT(3),
    SHORT(4),
    BYTE(5),
    ENUM(6),
    PACKET(7),
    OBJECT(8),
    NULL(9);

    public final int id;

    PType(int id) {
        this.id = id;
    }

    public static PType of(Class<?> clazz) {
        if (clazz == String.class)
            return STRING;
        if (clazz == Float.TYPE || clazz == Float.class || clazz == Double.TYPE || clazz == Double.class)
            return DOUBLE;
        if (clazz == Long.TYPE || clazz == Long.class)
            return LONG;
        if (clazz == Integer.TYPE || clazz == Integer.class)
            return INT;
        if (clazz == Short.TYPE || clazz == Short.class || clazz == Character.TYPE || clazz == Character.class)
            return SHORT;
        if (clazz == Boolean.TYPE || clazz == Boolean.class || clazz == Byte.TYPE || clazz == Byte.class)
            return BYTE;
        if (Enum.class.isAssignableFrom(clazz))
            return ENUM;
        if (Packet.class.isAssignableFrom(clazz))
            return PACKET;
        if (clazz == Void.TYPE || clazz == Void.class)
            return NULL;
        return OBJECT;
    }
}

File: src/main/kotlin/ru/kuris/PType.kt
package ru.kuris

import kotlin.reflect.KClass

enum class PType(id: Int) {
    STRING(0),
    DOUBLE(1),
    LONG(2),
    INT(3),
    SHORT(4),
    BYTE(5),
    ENUM(6),
    PACKET(7),
    OBJECT(8),
    NULL(9);

    companion object {
        @JvmStatic
        fun of(clazz: KClass<out Any>): PType {
            return when (clazz) {
                String::class -> STRING
                Double::class, Float::class -> DOUBLE
                Long::class -> LONG
                Int::class -> INT
                Short::class -> SHORT
                Byte::class -> BYTE
                Unit::class, Void::class -> NULL
                else -> {
                    if (Enum::class.java.isInstance(clazz.java))
                        ENUM
                    if (Packet::class.java.isInstance(clazz.java))
                        PACKET
                    OBJECT
                }
            }
        }
    }
}